### PR TITLE
Decode ClientState from Protobuf's Any type

### PR DESCRIPTION
**Pull request by romac**
See: #218

## Description

Decode both Tendermint and Mock client states from Protobuf's Any type.

______

For contributor use:

- [ ] Unit tests written
- [ ] Added test to CI if applicable 
- [ ] Updated CHANGELOG_PENDING.md
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: modules/src/ics02_client/client_def.rs [hunk 1]
```diff
@@ -1,24 +1,30 @@
...
 
 use ::tendermint::block::Height;
 
 #[cfg(test)]
-use crate::mock_client::client_def::MockClient;
-#[cfg(test)]
-use crate::mock_client::header::MockHeader;
-#[cfg(test)]
-use crate::mock_client::state::{MockClientState, MockConsensusState};
+use {
```
Comment (adizere) [inline modules/src/ics02_client/client_def.rs:22]: Ahh this is such a treat to the eyes.

File: modules/src/ics02_client/client_def.rs [hunk 2]
```diff
@@ -106,12 +113,39 @@ impl Header for AnyHeader {
...
+    Tendermint(TendermintClientState),
 
     #[cfg(test)]
     Mock(MockClientState),
 }
 
+impl AnyClientState {
+    pub fn from_any(any: prost_types::Any) -> Result<Self, Error> {
+        match any.type_url.as_str() {
+            "ibc.tendermint.ClientState" => {
```
Comment (adizere) [inline modules/src/ics02_client/client_def.rs:125]: Minor:

Maybe we should consider keeping these strings (lines 125 `"ibc.tendermint.ClientState"` and 135 `"ibc.mock.ClientState"`) hardcoded in some enum as opposed to hardcoded here? I can imagine they could even go in the ibc-proto repo, not here, but not sure what would be best.

To keep things simpler and get this merged fast, we can also just keep the hardcoded values as they are and just add a todo.

File: modules/src/ics03_connection/msgs.rs [hunk 3]
```diff
@@ -14,23 +14,26 @@
 //! TODO: Separate the Cosmos-SDK specific functionality from canonical ICS types. Decorators?
 
 #![allow(clippy::too_many_arguments)]
```
Comment (adizere) [inline modules/src/ics03_connection/msgs.rs:16]: Looking at the code, I think it's safe to remove this now.

File: modules/src/ics23_commitment/merkle.rs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 use crate::ics23_commitment::commitment::CommitmentPrefix;
-use ibc_proto::commitment::MerklePath;
+
+use ibc_proto::ibc::commitment::MerklePath;
```
Comment (adizere) [inline modules/src/ics23_commitment/merkle.rs:3]: Seems we always import types from ibc proto as `use ibc_proto::ibc::X`, so I'm wondering why is the `ibc` namespace in the middle necessary. We could just say `use ibc_proto::X`.

**Review by adizere — APPROVED** [review]
Looks great! I only had relatively minor comments.

File: modules/src/ics02_client/client_def.rs (hunk 2 above)
Comment (ancazamfir) [inline modules/src/ics02_client/client_def.rs:125]: Wondering if we should define these `type_url` strings in a separate place so they can also be used for encoding. We can do it later also.

**Review by ancazamfir — APPROVED** [review]
Looks great! thanks!

File: modules/src/ics02_client/client_def.rs (hunk 2 above)
Comment (romac) [reply modules/src/ics02_client/client_def.rs:125]: Yeah good point! Those should probably be defined as constants somewhere.

File: modules/src/ics23_commitment/merkle.rs (hunk 4 above)
Comment (romac) [reply modules/src/ics23_commitment/merkle.rs:3]: See this PR for a (poorly explained) rationale: https://github.com/informalsystems/ibc-proto/pull/9

Comment (romac) [reply modules/src/ics23_commitment/merkle.rs:3]: Also, the ibc_proto crate exports other types which are not ibc-specific so imho it makes sense to qualify their usage, though we could te-export them at the top level but that may cause similar issues to the one mentioned in the PR linked above.

File: modules/src/ics02_client/client_def.rs (hunk 2 above)
Comment (romac) [reply modules/src/ics02_client/client_def.rs:125]: Agreed! See my reply to Adi’s comment above.

**Pull request merged**